wdl-lint/src/rules/shellcheck.rs

```diff
@@ -375,13 +378,98 @@ impl<'a> CommandContext<'a> {
     }
 }
 
+/// Detect embedded quotes surrounding an expression in a string.
+///
+/// This is a utility function called by `evaluates_to_literal`. Only `expr`
+/// that are addition or strings with potentially embedded placeholders are
+/// valid input. For a given expression, it checks through all descendants to
+/// see if there are any name references (variables) that are surrounded by
+/// escaped quotes. In WDL, the parent expression is either an addition
+/// (concatenation, e.g. `~{"foo " + bar + " baz"}`) operation or a string with
+/// an embedded placeholder (e.g. `~{"foo ~{bar} baz"`). So the escaped quotes
+/// are not in a single string literal. The descendant expressions must be
+/// traversed to check for quoting.
+fn is_quoted(expr: &Expr) -> bool {
+    let mut opened = false;
+    let mut name = false;
+
+    for c in expr.descendants::<Expr>() {
+        match c {
+            Expr::Literal(LiteralExpr::String(ref s)) => {
+                for p in s.parts() {
+                    match p {
+                        StringPart::Text(t) => {
+                            let mut buffer = String::new();
+                            t.unescape_to(&mut buffer);
+                            buffer.match_indices(&['\'', '"']).for_each(|(..)| {
+                                if opened && name {
+                                    name = false;
+                                }
+                                opened = !opened;
+                            });
+                        }
+                        StringPart::Placeholder(_placeholder) => {
+                            if !opened {
+                                return false;
+                            }
+                            name = true;
+                        }
+                    }
+                }
+            }
+            Expr::NameRef(_) => {
+                if !opened {
+                    return false;
+                }
+                name = true;
+            }
+            _ => {}
+        }
+    }
+    !name
+}
+
+/// Evaluate an expression to determine if it can be simplified to a literal.
+fn evaluates_to_literal(expr: &Expr) -> bool {
+    match expr {
+        Expr::Literal(LiteralExpr::String(s)) => {
+            if s.text().is_some() {
+                return true;
+            }
+            is_quoted(expr)
+        }
+        Expr::Literal(_) => true,
+        Expr::Call(c) => match c.target().text() {
+            "sep" => evaluates_to_literal(
+                &c.arguments()
+                    .nth(1)
+                    .expect("`sep` call should have two arguments"),
+            ),
+            "quote" | "squote" => true,
+            _ => false,
+        },
```

Comment:
A comment explaining the criteria for inclusion here would be helpful. Two things in particular are making it hard for me to figure out what the rules are:

1. The return types of `quote` and `squote` are `Array[String]`, but `sep` returns `String`.
2. `prefix` and `suffix` seem quote similar to the other functions, but are left out.

Reply:
Definitely a good catch. We actually have two examples in our codebase that are now correctly handled.